enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,425 @@
+// referenced object inside the same namespace.
+// +structType=atomic
+type LocalObjectReference struct {
+	// name of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	// +optional
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+}
+```
+
+_Note_: The default value would be `nil`, both `nil` and empty `LocalObjectReference{}`
+are treated as though the field is unset. The secret is required to be created in the
+same namespace as that of the Route. The secret must be of type `kubernetes.io/tls` and
+the tls.key and the tls.crt key must be provided in the `data` (or `stringData`) field
+of the Secret configuration.
+
+If neither `.spec.tls.externalCertificate` or `.spec.tls.certificate` and `.spec.tls.key` are
+provided the router will serve the default generated certificates. User's will not be able to
+provide both `.spec.tls.certificate/key` and `.spec.tls.externalCertificate`. API server
+admission validation will enforce this.
+
+All valid and invalid scenarios will be depicted via the existing `RouteIngressCondition`.
+
+#### Variation
+
+N.A
+
+### Implementation Details/Notes/Constraints [optional]
+
+The router will read the secret referenced in `.spec.tls.externalCertificate` and will use
+the certificate inside to configure HAProxy if the secret is present and if the
+following pre-conditions are met:
+
+- Validations done by API server as part of [ValidateRoute()](https://github.com/openshift/openshift-apiserver/blob/aac3dd5bf0547e928103a0f718ca104b1bb13930/pkg/route/apis/route/validation/validation.go#L21),
+
+  - The router serviceaccount must have permission to read this secret particular secret.
+    - The role and rolebinding to provide this access must be provided by the user.
+  - CEL validations and openshfit/library-go will enforce that both `.spec.tls.certificate` and `.spec.tls.externalCertificate`
+    are not specified on the route at the same time. Since CEL validations are not run on Openshift API server, the same
+    validation will be done as part of `ValidationRoute()`.
+
+- New validation added to the API server as `ValidationHostExternalCertificate()`
+
+  - Any route that is updated or created which has a non-empty `.spec.tls.externalCertificate`,
+    will need require additional permission checks done as changing the certificate also affects
+    `.spec.host` or `.spec.subdomain`. Meaning any user that is updating the certificate must also
+    have `create` and `update` permission on the `custom-host` sub-resource.
+  - Any user that does not have both these permissions will not be allowed to update/create routes
+    that use `.spec.tls.externalCertificate`.
+  - This validation is intentionally done only on the API server and not the router as well, this
+    will result in not having this validation for events that are generated by the secret monitor
+    directly to the route controller in the router.
```

Comment:
This is redundant to what you said in the drawbacks. We can omit it or reference the drawback section.